src/main.cpp

```diff
@@ -2138,8 +2138,19 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
         }
     }
 
-    // set the old best anchor back
-    view.PopAnchor(blockUndo.old_tree_root);
+    // set the old best Sprout anchor back
+    view.PopAnchor(blockUndo.old_sprout_tree_root, SPROUT);
+
+    // set the old best Sapling anchor back
+    // We can get this from the `hashFinalSaplingRoot` of the last block
+    // However, this is only reliable if the last block was on or after
+    // the Sapling activation height. Otherwise, the last anchor was the
+    // empty root.
+    if (NetworkUpgradeActive(pindex->pprev->nHeight, Params().GetConsensus(), Consensus::UPGRADE_SAPLING)) {
+        view.PopAnchor(pindex->pprev->hashFinalSaplingRoot, SAPLING);
+    } else {
+        view.PopAnchor(ZCSaplingIncrementalMerkleTree::empty_root(), SAPLING);
```

Comment:
Does it make sense to pop an empty anchor repeatedly?

Reply:
Yep.

Reply:
Oh right, `PopAnchor(rt)` sets the best anchor to `rt` from whatever it currently is, rather than popping `rt` itself.

Nit: `RevertAnchorTo(rt)` or `SetAnchorTo(rt)` would be clearer.